README.md
Raspi PWM
==========

**NOTICE: This package has been deprecated in favor of [raspi-soft-pwm](https://github.com/nebrius/raspi-soft-pwm)**

[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/nebrius/raspi-io?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

Raspi PWM is part of the [Raspi.js suite](https://github.com/nebrius/raspi) that provides access to the hardware PWM.

If you have a bug report, feature request, or wish to contribute code, please be sure to check out the [Raspi IO Contributing Guide](https://github.com/nebrius/raspi-io/blob/master/CONTRIBUTING.md).

## System Requirements

- Raspberry Pi Model B Rev 1 or newer (sorry Model A users)
- Raspbian Jessie or newer
  - [Node-RED](http://nodered.org/) works, but can be finicky and difficult to debug.
  - See https://github.com/nebrius/raspi-io/issues/24 for more info about support for other OSes
- Node 6.4.0 or newer

Detailed instructions for getting a Raspberry Pi ready for NodeBots, including how to install Node.js, can be found in the [wiki](https://github.com/nebrius/raspi-io/wiki/Getting-a-Raspberry-Pi-ready-for-NodeBots)

## Installation

First, be sure that you have installed [raspi](https://github.com/nebrius/raspi).

Install with npm:

```sh
npm install raspi-pwm
```

**Note:** this project is written in [TypeScript](http://www.typescriptlang.org/) and includes type definitions in the package.json file. This means that if you want to use it from TypeScript, you don't need to install a separate @types module.

## Example Usage

In TypeScript/ES6:

```TypeScript
import { init } from 'raspi';
import { PWM } from 'raspi-pwm';

init(() => {
  const led = new PWM('P1-12');
  led.write(0.5); // 50% Duty Cycle, half brightness
});
```

In JavaScript:

```JavaScript
const raspi = require('raspi');
const pwm = require('raspi-pwm');

raspi.init(() => {
  const led = new pwm.PWM('P1-12');
  led.write(0.5); // 50% Duty Cycle, aka half brightness
});
```

## Pin Naming

The pins on the Raspberry Pi are a little complicated. There are multiple headers on some Raspberry Pis with extra pins, and the pin numbers are not consistent between Raspberry Pi board versions.

To help make it easier, you can specify pins in three ways. The first is to specify the pin by function, e.g. `'GPIO18'`. The second way is to specify by pin number, which is specified in the form "P[header]-[pin]", e.g. `'P1-7'`. The final way is specify the [Wiring Pi virtual pin number](http://wiringpi.com/pins/), e.g. `7`. If you specify a number instead of a string, it is assumed to be a Wiring Pi number.

Be sure to read the [full list of pins](https://github.com/nebrius/raspi-io/wiki/Pin-Information) on the supported models of the Raspberry Pi.

## PWM ports

Raspberry Pi Model B Rev 2 and older versions of the Raspberry Pi only expose one PWM port on pin `P1-12`, so be sure to always use that port (or leave it as the default). The Raspberry Pi Model B+ and newer exposes two PWM ports.

On newer Raspberry Pi's, these two PWM ports can be accessed via four different pins. `PWM0` can be accessed on either pin `P1-12` _or_ `P1-32`, and `PWM1` can be accessed on either pin `P1-33` _or_ `P1-35`. Each PWM _cannot_ be accessed on both pins though.

If you try to run the following code, which initializes both PWM0 ports,  _you will get an exception_.

```JavaScript
const raspi = require('raspi');
const pwm = require('raspi-pwm');

raspi.init(() => {
  const led1 = new pwm.PWM('P1-12');
  const led2 = new pwm.PWM('P1-32');
});
```

## API

### new PWM(config)

Instantiates a new PWM instance on the given pin. Note that PWM is limited to only 1 pin on the Model A/B and 2 pins on the A+/B+/2/Zero. On the A/B, the PWM pin is exposed on `GPIO18`. The A+/B+/2 is a little more complicated. The first PWM is exposed on two pins, `GPIO18` and `GPIO12`. This PWM can only be active on one of these pins at a time, however, so choose carefully. The second PWM pin on the A+/B+/2/Zero is exposed on `GPIO19`. Check the [wiring information wiki](https://github.com/nebrius/raspi-io/wiki) for more information.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>config (optional)</td>
    <td>Number | String | Object</td>
    <td>The configuration for the PWM pin. If the config is a number or string, it is assumed to be the pin number for the peripheral. If it is an object, the following properties are supported:</td>
  </tr>
  <tr>
    <td></td>
    <td colspan="2">
      <table>
        <thead>
          <tr>
            <th>Property</th>
            <th>Type</th>
            <th>Description</th>
          </tr>
        </thead>
        <tr>
          <td>pin (optional)</td>
          <td>Number | String</td>
          <td>The pin number or descriptor for the peripheral. Defaults to 1 (GPIO18, PWM0).</td>
        </tr>
        <tr>
          <td>frequency (optional)</td>
          <td>Number</td>
          <td>The frequency, in Hz, of the PWM signal. Defaults to 50.</td>
        </tr>
      </table>
    </td>
  </tr>
</table>

### Instance Properties

#### frequency

A number representing the frequency initialization value, in Hz. If a value for `frequency` was passed to the constructor, it is reflected back here. If no value for `frequency` was passed to the constructor, then this reflects the default frequency value of `50`.

#### dutyCycle

A number representing the last written dutyCycle. If this property is read before a value is written, `0` is returned.

### Instance Methods

#### write(dutyCycle)

Sets the duty cycle for the PWM output, a floating point value between 0 and 1.

_Arguments_:

<table>
  <thead>
    <tr>
      <th>Argument</th>
      <th>Type</th>
      <th>Description</th>
    </tr>
  </thead>
  <tr>
    <td>dutyCycle</td>
    <td>Number</td>
    <td>The duty cycle for the PWM to set, must be a floating point number between 0 and 1</td>
  </tr>
</table>

_Returns_: None

**Note:** The PWM does not start outputting a signal until write is called for the first time.

License
=======

The MIT License (MIT)

Copyright (c) 2014-2017 Bryan Hughes <bryan@nebri.us>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.


CHANGELOG.md
## 6.0.1 (2021-10-11)

- Deprecated this package in favor of [raspi-soft-pwm](https://github.com/nebrius/raspi-soft-pwm)

## 6.0.0 (2018-4-21)

- Updated pigpio to v1
- BREAKING CHANGE: dropped support for Node.js 4

## 5.0.0 (2017-9-12)

- POTENTIALLY BREAKING CHANGE: renamed the following private properties to begin with an underscore because I just learned TypeScript doesn't do anything to hide private properties. Don't use these properties, changes may not be semver-major next time.
    - `PWM.frequencyValue` -> `PWM._frequencyValue`
    - `PWM.dutyCycleValue` -> `PWM._dutyCycleValue`
    - `PWM.pwmPort` -> `PWM._pwmPort`
    - `PWM.pwm` -> `PWM._pwm`
- BREAKING CHANGE: dropped support for Node.js < 4.0.0, and now enforce it via package.json "engines" field.
- SORT OF BREAKING-ISH CHANGE: dropped support for attempting to install on non-arm platforms via package.json's "cpu" field.
    - Attempting to install this on a non-Raspberry Pi platform before gave a bunch of obtuse errors, so this doesn't _actually_ change the ability to install raspi-pwm, but does make it fail earlier and harder.

## 4.0.2 (2017-4-25)

- Fixed a bug in the TypeScript definitions that said `frequencey` was required when in reality it's not.

## 4.0.1 (2017-4-23)

- Updated documentation

## 4.0.0 (2017-4-23)

- Removed Wiring Pi and replaced it with pigpio.
- BREAKING CHANGE: removed the `range` and `clockDivisor` constructor options and instance properties
- Added `frequency` constructor option and instance property. Replace the use of `range` and `clockDivisor` with this property.
- BREAKING CHANGE: `write` now takes a floating point number between 0 and 1, instead of an integer between 0 and 1024.

## 3.1.3 (2017-22-1)

- Publishing a new version to update the README on npmjs.com. No other changes.

## 3.1.2 (2016-12-21)

- Fixed a TypeScript API definition bug where the constructor config was not marked as optional
  - Note: this does not affect functionality unless you were using the TypeScript definition

## 3.1.1 (2016-12-3)

- Converted the project to TypeScript and cleaned up a bunch of odds and ends
  - Note: there is no functionality change or bug fixes with this release

## 3.1.0 (2016-7-7)

- Switched dependency ranges to ^
- Bumped dependencies to bring in support for a new Raspberry Pi Zero revision

## 3.0.0 (2016-3-20)

- Dependency update to fix bug
  - POTENTIALLY BREAKING CHANGE: using the identifier `PWM1` now maps to Wiring Pi pin 23, not 24
- New build system

## 2.3.1 (2016-3-7)

- Dependency update to add missing Raspberry Pi 3 Model B revision

## 2.3.0 (2016-3-4)

- Updated dependencies to add Raspberry Pi 3 Model B support

## 2.2.0 (2016-2-18)

- Added `range` and `clockDivisor` accessors

## 2.1.0 (2015-12-8)

- Updated dependencies to add Raspberry Pi Zero support

## 2.0.0 (2015-10-20)

- Upgraded to NAN 2
  - POTENTIAL BREAKING CHANGE
  - The API has not changed, but the build requirements have
  - Make sure you are running Raspbian Jessie because this module no longer builds on stock Raspbian Wheezy
  - See https://github.com/fivdi/onoff/wiki/Node.js-v4-and-native-addons for more information

## 1.4.0 (2015-10-13)

- Exposed clock and range properties
- Dependency updates to fix bug with invalid pin aliases
- Updated build dependencies

## 1.3.3 (2015-9-3)

- Dependency updates to fix a bug with pin aliasing

## 1.3.2 (2015-7-16)

- Updated dependencies
- Updated the repository links to point to their new location
- Added a contributing guide
- Added code linter
- Update code style to use newer best practices

## 1.3.1 (2015-3-17)

- Dependency update to fix a bug with destroying peripherals

## 1.3.0 (2015-2-21)

- Switched from traceur to babel for ES6->ES5 compilation

## 1.2.2 (2015-2-20)

- Fixed a bug with mismatched versions of the traceur runtime
- Skipped 1.2.1 because of a typo on my part when publishing (oops)

## 1.2.0 (2015-2-20)

- Removed prebuilt binary support as there are now multiple targets (need to use node-pre-gyp)
- Upgraded NAN to get support for Node.js 0.12
  - io.js support is theoretically there, but won't work until https://github.com/TooTallNate/node-gyp/pull/564 is landed

## 1.1.0 (2015-2-13)

- Added support for multiple PWMs. Note: the API was extended, but is backwards compatible with 1.0.x releases

## 1.0.5 (2015-2-12)

- README updates

## 1.0.4 (2015-1-21)

- Locked down the NAN version for now since code breaks on 1.5

## 1.0.3 (2015-1-8)

- Added pre-built lib so users don't have to compile on their own Raspberry Pis.

## 1.0.2 (2014-12-29)

- Added docs
- Switched to using the 'PWM0' pin nomenclature

## 1.0.1 (2014-12-6)

- Critical bug fix (why did I even publish 1.0.0?)

## 1.0.0 (2014-11-28)

- Initial implementation


src/index.ts
/*
The MIT License (MIT)

Copyright (c) 2014-2017 Bryan Hughes <bryan@nebri.us>

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
*/

import { Peripheral } from 'raspi-peripheral';
import { Gpio } from 'pigpio';

export interface IConfig {
  pin?: number | string;
  frequency?: number;
}

const DEFAULT_PIN = 1;
const DEFAULT_FREQUENCY = 50;
const MAX_DUTY_CYCLE = 1000000;

const PWM0 = 'PWM0';
const PWM1 = 'PWM1';

// So there's a funky thing with PWM, where there are four PWM-capable pins,
// but only two actual PWM ports. So the standard pin contention mechanism
// doesn't _quite_ cover all cases. This object tracks which PWM peripherals are
// in use at a given time, so we can do error checking on it.
const pwmPeripheralsInUse: { [ port: string ]: boolean } = {
  [PWM0]: false,
  [PWM1]: false
};

export class PWM extends Peripheral {

  private _frequencyValue: number;
  private _dutyCycleValue: number;
  private _pwmPort: string;

  private _pwm: Gpio;

  public get frequency() {
    return this._frequencyValue;
  }

  public get dutyCycle() {
    return this._dutyCycleValue;
  }

  constructor(config?: number | string | IConfig) {
    let pin: number | string = DEFAULT_PIN;
    let frequency = DEFAULT_FREQUENCY;
    if (typeof config === 'number' || typeof config === 'string') {
      pin = config;
    } else if (typeof config === 'object') {
      if (typeof config.pin === 'number' || typeof config.pin === 'string') {
        pin = config.pin;
      }
      if (typeof config.frequency === 'number') {
        frequency = config.frequency;
      }
    }
    super(pin);

    // Pin details from http://elinux.org/RPi_BCM2835_GPIOs
    let gpioPin: number;
    let mode: number;
    switch (this.pins[0]) {
      case 26: // GPIO12 PWM0 ALT0
        gpioPin = 12;
        mode = Gpio.ALT0;
        this._pwmPort = PWM0;
        break;
      case 1: // GPIO18 PWM0 ALT5
        gpioPin = 18;
        mode = Gpio.ALT5;
        this._pwmPort = PWM0;
        break;
      case 23: // GPIO13 PWM1 ALT0
        gpioPin = 13;
        mode = Gpio.ALT0;
        this._pwmPort = PWM1;
        break;
      case 24: // GPIO19 PWM1 ALT5
        gpioPin = 19;
        mode = Gpio.ALT5;
        this._pwmPort = PWM1;
        break;
      default:
        throw new Error(`Pin ${pin} does not support hardware PWM`);
    }

    if (pwmPeripheralsInUse[this._pwmPort]) {
      throw new Error(`${this._pwmPort} is already in use and cannot be used again`);
    }
    pwmPeripheralsInUse[this._pwmPort] = true;

    this._frequencyValue = frequency;
    this._dutyCycleValue = 0;
    this._pwm = new Gpio(gpioPin, { mode });
  }

  public destroy() {
    pwmPeripheralsInUse[this._pwmPort] = false;
    super.destroy();
  }

  public write(dutyCycle: number) {
    if (!this.alive) {
      throw new Error('Attempted to write to a destroyed peripheral');
    }
    if (typeof dutyCycle !== 'number' || dutyCycle < 0 || dutyCycle > 1) {
      throw new Error(`Invalid PWM duty cycle ${dutyCycle}`);
    }
    this._dutyCycleValue = dutyCycle;
    this._pwm.hardwarePwmWrite(this._frequencyValue, Math.round(this._dutyCycleValue * MAX_DUTY_CYCLE));
    this.emit('change', this._dutyCycleValue);
  }
}
